### PR TITLE
refactor(core): demote validate-plan from public api

### PR DESCRIPTION
**Pull request by christopher-buss**
## Summary

- Publishes `packages/bedrock/CONTEXT.md` for `@bedrock-rbx/core`: an 18-term glossary across Resources, State, Reconciliation, Plugin contracts, and Patterns. The grilling session that produced it pinned the canonical vocabulary (Resource, Kind, Key, Desired/Current state, Outputs, Operation, Diff, Apply, Deploy, Reconcile, Drift, Driver, Kind module, State port, Progress port, Backend, Environment, Adoption, Managed field, Redaction).
- Demotes `validatePlan` from the public API. The function only exists to dispatch to each kind module's `assertReconcilable` hook; there was no real consumer use case for calling it standalone. Internal callers (`deploy`, `previewDiff`, integration tests) now reach it via the internal path. **Breaking change for any direct consumer of the export**, but bedrock is pre-1.0 and the function had no documented standalone use.
- Renames `validatePlan` -> `assertAllReconcilable` so the name reflects what it does (asserts reconcilability for every paired entry) and mirrors the per-kind `ResourceKindModule.assertReconcilable` hook it dispatches to.
- Aligns JSDoc across 6 files to the glossary: drops "plan-time", "plan output", "Plan and apply renderers", "validatePlan saw" in favor of Diff / preview / reconcilability / `assertAllReconcilable`. Drops "Last-known live state" -> "Last-known current state" in `operations.ts` and `diff.ts`; the CLI's "Reconciles config to live state" becomes "Reconciles config to upstream state".
- Bumps three agent skills (`grill-with-docs` format simplification, `handoff` tweaks, `setup-matt-pocock-skills/issue-tracker-gitlab` note removal) and adds the new `prototype` skill. Tracked separately in `chore: bump agent skills versions`.

## Test plan

- [x] `pnpm lint` clean
- [x] `pnpm test` green (the renamed `assert-all-reconcilable.spec.ts` carries every prior `validate-plan.spec.ts` case verbatim under the new name)
- [x] `pnpm typecheck` clean
- [x] `pnpm build` succeeds
- [x] `pnpm mutate:changed` clean (no surviving mutants on touched `src/**`)
- [x] Integration test `tests/integration/products-redacted.spec.ts` imports `assertAllReconcilable` via `#src/core/assert-all-reconcilable` (internal path) and still exercises the redacted-product flow end-to-end

## Notes for reviewers

- The CONTEXT.md format was simplified mid-session via a skill update: Relationships, Example dialogue, and Flagged ambiguities sections are no longer part of the structure. The committed glossary already matches the new format.
- Two parallel review subagents (terminology + coverage) sharpened the glossary before this alignment pass landed. They surfaced Managed field and Progress port as missing terms (now included) and caught Backend over-generalising to "any plugin contract" when only the State port is backend-configurable today.
- `planFollowUpPatch` in `src/core/plan-follow-up-patch.ts` is deliberately untouched: "plan" there is a verb meaning "derive/compute the necessary X", not the IaC noun.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

# Bedrock Code Review: PR `#440` - Demote validatePlan from Public API

## ✅ Architecture Compliance

**FCIS Pattern**: Excellent adherence
- **Core Layer**: `assertAllReconcilable` is a pure function with no I/O or side effects. It performs collision detection via `detectProductNameCollision` and dispatches to kind modules' optional `assertReconcilable` hooks, returning typed `Result` errors. Properly located in `packages/bedrock/src/core/`.
- **Shell Layer**: Correct usage in `deploy.ts` and `preview-diff.ts` — orchestration only, delegating validation to Core before `diff` runs.
- **No cross-layer violations**: Function does not perform I/O or business logic outside Core.

---

## ✅ Testing Compliance (TDD Required)

**Test Coverage**: Comprehensive
- `assert-all-reconcilable.spec.ts`: 192 lines covering:
  - ✅ Success cases (no current state, valid icon transitions, composite key matching)
  - ✅ Icon removal rejection when current has icon and desired drops it
  - ✅ Developer-product name collision detection (both hashed and user-overridden names)
  - ✅ Orphaned current entries (correctly ignored)
  - ✅ Kind+Key composite pairing logic
  - ✅ Per-kind uniqueness scoping (developer-product names don't collide with game-pass names)

**Integration Tests**: End-to-end validation in `products-redacted.spec.ts`
- ✅ Two tests updated to use `assertAllReconcilable` via internal path `#src/core/assert-all-reconcilable`
- ✅ Tests exercise redacted-product flow with distinct wire names and collision rejection
- ✅ Confirms function gates invalid plans before `diff` and apply-side driver I/O

**Mutation Testing**: PR notes `pnpm mutate:changed` passed with no surviving mutants on touched src/**

---

## ✅ Test Isolation & Structure

- **Core tests**: Unit tests with no mocking needed (pure synchronous function)
- **Integration tests**: Use fake drivers and HTTP mocks (nock pattern via `createFakeHttpClient`)
- **Naming conventions**: Proper `it("should <behavior>")` format observed throughout

---

## ✅ API Design & Breaking Changes

**Demotion Complete**:
- Removed from `packages/bedrock/src/index.ts` (public re-export)
- Renamed `validatePlan` → `assertAllReconcilable` (clearer intent: asserts reconcilability for all paired entries)
- Still exported from `./core/assert-all-reconcilable.ts` but accessed via internal path by callers
- Integration tests import via `#src/core/assert-all-reconcilable` (internal alias), confirming public API is broken as intended
- Pre-1.0 package: documented breaking change acceptable

**All Usage Sites Migrated**:
- ✅ `deploy.ts`: switched to `assertAllReconcilable` import
- ✅ `preview-diff.ts`: switched to `assertAllReconcilable` import, wires failure into existing `{ kind: "buildDesiredFailed" }` error shape
- ✅ `products-redacted.spec.ts`: integration tests updated with internal import path
- ✅ Removed `validate-plan.example.spec.ts` (stale example removed)

---

## ✅ Error Handling & Types

- **Typed errors**: Returns `Result<undefined, BuildDesiredError>` with discriminated union kinds:
  - `iconRemovalRejected`: Kind-level invariant (current has icon, desired drops it)
  - `redactedNameCollision`: Batch-level invariant (two developer-products resolve to same wire name)
- **Error messages**: Include resource keys and collision details for debugging
- **No unhandled cases**: All code paths return or throw explicitly typed errors

---

## ✅ Documentation & Vocabulary

**CONTEXT.md**: Well-structured glossary added covering:
- Resources, Kind, Key
- Desired/Current state, Outputs
- State, Environment
- Operations, Diff, Apply, Deploy, Reconcile, Drift
- Driver, Kind module, State/Progress ports, Backend
- Adoption, Managed field, Redaction patterns

**JSDoc Updates** across 6 files (aligned to new glossary):
- `cli/index.ts`: "upstream state" instead of "live state"
- `core/diff.ts`: "Last-known current state" instead of "Last-known live state"
- `core/kinds/developer-product.ts`: "pre-reconcile" instead of "plan-time"
- `core/kinds/module.ts`: `assertReconcilable` documented as "optional reconcilability invariant called by `assertAllReconcilable`"
- `core/operations.ts`: "preview and apply" renderers (instead of "plan and apply")
- `core/redact-resources.ts`: Updated references to detection phase

---

## ⚠️ Minor Observations

1. **Module reference in pure function**: `assertAllReconcilable` uses module-scoped `defaultKindRegistry` directly rather than injecting it as a dependency. This is acceptable for pure functions but could be noted as intentional design (registry is stable, not a side effect).

2. **Kind module discovery**: Function assumes all kind modules exist in `defaultKindRegistry` and will assert if missing. Edge case coverage is solid (all kinds with optional hooks are tested).

3. **Composite key format** (`kind:key`): Implementation detail is sound, tests confirm pairing logic works as intended.

---

## Summary

**Status**: ✅ **APPROVED** for merge

This PR successfully demotes `validatePlan` from public API with proper renaming, comprehensive test coverage, end-to-end integration validation, and consistent documentation updates. The Core function is pure, the Shell layer orchestrates correctly, error handling is typed and complete, and all usage sites have been migrated. Breaking change is appropriate for pre-1.0 package with no documented standalone use of the demoted function.

[Review Change Stack](https://app.coderabbit.ai/change-stack/christopher-buss/bedrock/pull/440?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: packages/bedrock/src/core/kinds/module.ts [hunk 1]
```diff
@@ -12,11 +12,11 @@ import type { ResolvedConfig, ResourceEntryByKind } from "../schema.ts";
  *
  * - `fileReadFailed`: a kind module's `normalize` could not read a file
  *   the input declared (e.g. An icon path that is missing on disk).
- * - `iconRemovalRejected`: `validatePlan` saw a kind whose prior current
- *   state recorded an icon that the desired state no longer declares,
+ * - `iconRemovalRejected`: `assertAllReconcilable` saw a kind whose prior
```
[the remark this answers is not in the document]
Comment (christopher-buss) [reply packages/bedrock/src/core/kinds/module.ts]: Reworded the umbrella to 'Failure surfaced before `diff` runs' so it covers all three variants. The per-variant blurbs now describe the trigger directly rather than naming the internal function.

File: packages/bedrock/tests/integration/products-redacted.spec.ts [hunk 2]
```diff
@@ -596,7 +596,7 @@ describe("products-redacted pipeline end-to-end", () => {
 		const desiredResult = await buildDesired(flattenConfig(resolved.data), readFile);
 		assert(desiredResult.success);
 
-		const planCheck = validatePlan(desiredResult.data, []);
+		const planCheck = assertAllReconcilable(desiredResult.data, []);
 		assert(planCheck.success);
```
[the remark this answers is not in the document]
Comment (christopher-buss) [reply packages/bedrock/tests/integration/products-redacted.spec.ts]: Renamed `planCheck` to `reconcilable` at both call sites (lines 600 and 684).
